### DISCUSSION TITLE
Validierung Codierung DID 2622, 2623, 2624, 2625

### DISCUSSION THREAD

**Discussion by bu-na** [open3e/open3e#157] [Q&A] [answered]
Benötige hinsichtlich korrekter Dekodierung etwas Hilfe, vielleicht hat jemand noch einen bessere Idee...

**2622** :  O3EComplexType(9, "SeasonalCoefficientOfPerformaceHeating", [O3EInt8(2, "Actual",  scale=10), O3EInt8(2, "Maximum", scale=10), O3EInt8(2, "Minimum", scale=10), O3EInt8(2, "Average", scale=10), O3EByteVal(1, "Unknown")]), #250-xH 
**2623** :  O3EComplexType(9, "SeasonalEnergyEfficiencyRatioCooling", [O3EInt8(2, "Actual", scale=10), O3EInt8(2, "Maximum", scale=10), O3EInt8(2, "Mimimum", scale=10), O3EInt8(2, "Average", scale=10), O3EByteVal(1, "Unknown")]), #250-xH 
**2624** :  O3EComplexType(9, "SeasonalCoefficientOfPerformaceDomesticHotWater", [O3EInt8(2, "Actual", scale=10), O3EInt8(2, "Maximum", scale=10), O3EInt8(2, "Minimum", scale=10), O3EInt8(2, "Average", scale=10), O3EByteVal(1, "Unknown")]), #250-xH 
**2625** :  O3EComplexType(9, "SeasonalCoefficientOfPerformaceHeatingAndDomesticHotWater", [O3EInt8(2, "Actual", scale=10), O3EInt8(2, "Maximum",scale=10), O3EInt8(2, "Minimum", scale=10), O3EInt8(2, "Average", scale=10), O3EByteVal(1, "Unknown")]), #250-xH 
 
Gruß Bu-Na

**Comment by Philip-Wiege** [comment]
Hi,

ich hab mir gerade mal die Werte bei mir angeschaut und mit der App verglichen. Bei den jeweils ersten 2 Bytes gehe ich mit. Das sind die in der App angezeigten Jahresarbeitszahlen als signed Integer mit Skalierung 10. Woher hast du die Info, dass die anderen Bytes ein Maximum und Minimum sind? Kann man das über Werte aus der App plausibilisieren?

Grüße
Philip

**Comment by bu-na** [comment]
Es ist eine Vermutung in Anlehnung an die Decodierung anderer 9stelliger DIDs (z.B. 1768-1772).
Scheinbar werden die Daten auch recht selten aktualisiert, daher ist die Interpretation nicht ganz einfach.

Deshalb auch die Frage in die Runde...

Gruß Bu-Na

**Comment by Philip-Wiege** [comment]
Hi @bu-na! Ich wollte nicht unhöflich rüberkommen. Die ersten 2 Bytes sind ja schonmal plausibel und das könnten wir ja einbauen. Im Zweifel ist der Rest halt "Unknown". Ich habe nur Daten seit März in ViCare. Hast du schon für mehr als ein Jahr Daten in deiner ViCare App? Berechnet Viessmann die Arbeitszahlen für die letzten 12 Monate oder für ein Kalenderjahr? Kannst du das rausfinden?

**Comment by bu-na** [comment]
Die Daten werden (vermutlich) zum Ersten des Monats (kurz nach Null Uhr) aktualisiert. 

Hinsichtlich der Datenstrukturen würde ich immer von einer gewissen Vereinheitlichung ausgehen. 
Ob alle Elemente regulär bedient werden und korrekte Werte enthalten, steht allerdings auf einem anderen Blatt. Es gibt diverse DIDs wo z.B. Actual und Average identische live-Werte liefern, aber Maximum und Minimum scheinbar unplausible Werte enthalten.

Mehr Informationen liegen mir zur Zeit nicht vor, es hilft nur, weiter Daten loggen und die Ergebnisse schrittweise interpretieren..

Gruß Bu-Na

**Comment by Philip-Wiege** [comment, marked as answer]
Hi,

ich habs gerade mal in den Develop Branch eingestellt. Die Werte für die Jahresdaten müssten sogar nur im ersten Byte kodiert werden. Das erste Byte muss auf einen 8bit Integer umgewandelt werden und dann mit Faktor 10 geteilt werden. So passt es zu den App Daten.

**Comment by Philip-Wiege** [comment]
Bei den Energiematrizen gibt es Werte für jährlich, monatlich, wöchentlich etc. Vielleicht werden diese DIDs auch so aufgebaut... Könnte sich lohnen das mal durchzurechnen. Vielleicht steht jedes Byte für eine "Arbeitszahl" Daher könnten auch die "00" Bytes kommen z.B fürs Kühlen, da ich zum Beispiel diesen Monat nicht gekühlt habe.

**Reply by DpunktS** [reply]
Bei mir setzen sich die DIDs so zusammen.
Wert: 4,5
Summe thermisch: 13450 kWh
Summe Verbrauch: 2949 kWh

**Reply by bu-na** [reply]
Wäre ein Variante, jedenfalls sind es keine Min-/Max-Werte. Falls meine Vermutung stimmt, sollte es zum Monatswechsel neue Daten geben. Vielleicht hilft uns das weiter (RAW-Logging nicht vergessen).
Immerhin, der Anfang ist mit der Dekodierung der aktuellen JAZ gemacht.
(Gut wäre auch, ein paar RAW-Werte zu haben...)

**Reply by Philip-Wiege** [reply]
@DpunktS Meinst du wir haben also 1 Byte für die JAZ / SEER und dann jeweils 4 Bytes für die thermische und elektrische Energie? Wie bist du an die Dekodierung gekommen?

**Reply by DpunktS** [reply]
@Philip-Wiege die Werte sollten so passen.
Ich hatte mich damit schonmal beschäftigt. :)

**Reply by Philip-Wiege** [reply]
Komischerweise Passt das bei mir nur beim Kühlen und liegt beim Heizen voll daneben. Da liegt irgendwie zwischen den realen Werten der App und den Open3E Werten kein ganzzahliger schöner Faktor...

Heizen
3f 5b470000 510b0000
roh: 63 18267 2897
app: 6.3 2910 470

Warmwasser
2c 45550000 1f130000
roh: 44 21829 4895
app: 4.4 2450 554

HZ WW
33 a09c0000 701e0000
roh: 51 40096 7792
app: 5.1 5360 1024

Kühlen
40 5c2c0000 df060000
roh: 64 11356 1759
app: 6.4 1114 176

**Reply by bu-na** [reply]
Vermute mal, die Werte beziehen sich auf die letzten 365 Tage.  
Daher passen die Werte für DID2623 (Cooling), bei DID2622,2624,2625 fehlt der Rest aus dem vergangenen Jahr. 
Zumindest bei mir würde das funktionieren.

Die Codierung könnte damit wie folgt aussehen.
**2622** : O3EComplexType(9, "SeasonalCoefficientOfPerformaceHeating", [O3EInt8(1, "Actual",  scale=10), O3EInt32(4,"GeneratedHeatingOutput365", scale=10), O3EInt32(4, "GeneratedEnergyConsumption365", scale=10), O3EByteVal(1, "Unknown")]), #250-xH Unit: kWh
**2623** : O3EComplexType(9, "SeasonalEnergyEfficiencyRatioCooling", [O3EInt8(1, "Actual",  scale=10), O3EInt32(4,"GeneratedCoolingOutput365", scale=10), O3EInt32(4, "EnergyConsumptionCooling365", scale=10), O3EByteVal(1, "Unknown")]), #250-xH Unit: kWh
**2624** : O3EComplexType(9, "SeasonalCoefficientOfPerformaceDomesticHotWater", [O3EInt8(1, "Actual",  scale=10), O3EInt32(4,"GeneratedDHWOutput365", scale=10), O3EInt32(4, "EnergyConsumptionDHW365", scale=10), O3EByteVal(1, "Unknown")]), #250-xH Unit: kWh
**2625** : O3EComplexType(9, "SeasonalCoefficientOfPerformaceHeatingAndDHW", [O3EInt8(1,"Actual", scale=10), O3EInt32(4, "GeneratedHeatAndDHWOutput365",  scale=10), O3EInt32(4,"EnergyConsumptionAndDHWOutput365",scale=10), O3EByteVal(1, "Unknown")]), #250-xH Unit: kWh



Gruß Bu-Na

**Reply by DpunktS** [reply]
@Philip-Wiege bei mir passen die Werte:

```
DID2622 2d980b0200ee710000
Wert: 4,5			45	/10
Summe thermisch: 13.404kWh	134040	/10
Summe Verbrauch: 2.917kWh	29166	/10

DID2623 000000000000000000
Wert: 0				0	/10
Summe thermisch: 0kWh		0	/10
Summe Verbrauch: 0kWh		0	/10

DID2624 2839ce000043330000
Wert: 4				40	/10
Summe thermisch: 5.279kWh	52793	/10
Summe Verbrauch: 1.312kWh	13123	/10


DID2625 2cd1d9020031a50000
Wert: 4,4			44	/10
Summe thermisch: 18.683kWh	186833	/10
Summe Verbrauch: 4.229kWh	42289	/10


```

**Reply by Philip-Wiege** [reply]
Dann sollte das ja passen auch wenn meine Anlage komische Werte ausgibt. Ich könnte das später in den Develop Branch packen. Ich würde auch einen kleinen Rechtschreibfehler korrigieren. Da steht glaube ich PerformaceRatio.

**Comment by Philip-Wiege** [comment]
Unsere Erkenntnisse sind jetzt in Develop und fließen dann mit der Zeit auch in Master!
Danke für eure Unterstützung!

https://github.com/open3e/open3e/commit/be0c19db0636b17288fa34265cefdbacf01f8eee
